Refine DB structure

42Ar opened the pull request:
Refined Material fields by specializing Material class. Dropped color table, using RAL color numbers from now on. Improved helptexts and translations.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -1,13 +1,29 @@
+from django.utils import translation
+
+
+ral_color_table_source = ('django-plastic-tickets/' +
+                          'django_plastic_tickets/ral_colors.json')
+with open(ral_color_table_source) as f:
+    ral_color_table = json.load(f)
+ral_colors = {row['RAL']: row for row in ral_color_table}
+
+
+def validate_ral_number(val):
+    if val not in ral_colors:
+        raise ValidationError(gettext('This is not a valid RAL classic color number'))
```
Can we have this in its own (static) class?

42Ar replied:
I will put it in a separate module.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -1,13 +1,29 @@
+
+def validate_ral_number(val):
+    if val not in ral_colors:
+        raise ValidationError(gettext('This is not a valid RAL classic color number'))
```
style: Line 26 is too long, please rewrap

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
+    url = models.URLField(blank=True, null=True)
 
 
 class Material(models.Model):
```
Shouldn't this be abstract?

42Ar replied:
To my understanding of this, if Material was abstract class then it would not appear in the database. Thus, we could not reference it from MaterialStock. We can make Material abstract, but then we would have to add a field to MaterialStock (one to reference MaterialFilament and one to reference MaterialResin) which would not be very nice. There are also other approaches, but using Multi-table inheritance here is the best option I think.

hd1ex replied:
Ok agreeing with that

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
+        return f'{self.name} ({self.type.name}, {self.get_color_name()})'
+
+
+class MaterialResin(Material):
+    pass
+
+
+class MaterialFilament(Material):
```
I think these names should be the other way around
-> `ResinMaterial` and `FilamentMaterial`

42Ar replied:
I wanted to have a common prefix for all tables which have something to do with Materials. I added human readable names for the admin fronted.

hd1ex replied:
I'd rather have the names to be consistent with the admin frontend.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
+
+
+class MaterialFilament(Material):
+    nozzle_temp = models.FloatField(blank=True, null=True)
+    bed_temp = models.FloatField(blank=True, null=True)
+    diameter = models.FloatField(blank=True, null=True, default=1.75)
```
These fields need descriptions and should be next to the other ones above the methods

I expect these are still recommended optimal temps?

42Ar replied:
Yes these are the (manufacturer) recommended settings.

hd1ex replied:
Ok. Then please add this info at least to the descriptions.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
+    manufacturer = models.ForeignKey(Manufacturer, on_delete=models.CASCADE)
+    empty_weight = models.FloatField(blank=True, null=True)
+    usual_capacity = models.FloatField(blank=True, null=True)
+    appearance = models.TextField()
```
Please also add descriptions

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
     def __str__(self) -> str:
-        return f'{self.name} ({self.type.name}, {self.color.name})'
+        return (f'{self.manufacturer}, ' +
+                f'{self.usual_capacity} g,i ' +
```
What does `g,i` mean?

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
         help_text=gettext('The internal label to identify the material'),
         unique=True)
-    consumed = models.BooleanField(default=False)
+    initial_material_weight = models.FloatField(help_text=gettext(
+        'Initial weight of the material in gram, ' +
+        'not including the weight of the package'))
+    current_weight = models.FloatField(help_text=gettext(
+        'Current weight of the material in gram, ' +
+        'including the weight of the package'))
```
Why one including and one not including the weight of the package?    
Wouldn't this confusing?

42Ar replied:
This might be confusing at the beginning, but you know the initial weight of the material on a filament roll or resin bottle, but you can only measure the weight of it with the package.

hd1ex replied:
I am agreeing that `current_weight` should contain the package weight but then you also have to store the **weight** of the **package** in some way, otherwise the `initial_material_weight` would solve no purpose.

I guess this was your intention with `MaterialPackage`?     
So would some inventory workflow be this:
- Add a `MaterialPackage` for a light, transparent filament spool (from Extrudr)
- Add a `Material` for brown GreenTEC from Extrudr
- For every new filament in stock create a `MaterialStock` with the weight written on the package as `initial_material_weight` and the `current_weight = empty_weight + initial_material_weight`
- Calculate the remaining weight at any time with `current_weight - empty_weight`
Can you ACK or update this example?

One side effect of this is, that we are assuming an equal weight of every package.

Another small thing I have seen is that both `Material` and `MaterialPackage` are storing a `Manufacturer`. Is this intentional?

42Ar replied:
- Yes, this is what the typical workflow would look like.
- I think it is safe to assume that every package which looks the same has also the same weight. One can still weight a new package with material and subtract the weight of the material (which is known from the manufacturer) to check that.
- It is intentional that both `Material` and `MaterialPackage` store `Manufacturer`. This helps to distinguish the packages.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
+    price = models.FloatField(help_text=gettext(
+        'Total price, including the package.'))
+    owner_identifier = models.TextField(
+        help_text=gettext('This should be a LDAP ID in most cases'))
```
We have the `User?-model for this purpose, please use it

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -117,10 +181,10 @@ def get_message_row_count(self, cols=77) -> int:
 
 class PrintConfig(models.Model):
-    file = models.FilePathField()
+    file = models.TextField()
```
Does this fix https://github.com/hd1ex/plastic-tickets/issues/1 ?

If so, please put this in another commit or, even better, in another pull request.

42Ar replied:
Yes it does. I will make a separate commit or pull request.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -117,10 +181,10 @@ def get_message_row_count(self, cols=77) -> int:
     count = models.PositiveIntegerField()
     material_type = models.ForeignKey(MaterialType, on_delete=models.CASCADE)
-    color = models.ForeignKey(MaterialColor, on_delete=models.CASCADE)
+    color = models.PositiveIntegerField()
```
Please add a description for this

hd1ex commented on django_plastic_tickets/ral_colors.json:
```diff
@@ -0,0 +1 @@
+[{"RAL": 1000, "RGB": "190-189-127", "HEX": "#BEBD7F", "de": "Gr\u00fcnbeige", "en": "Green beige", "fr": "Beige vert", "es": "Beige verdoso", "it": "Beige verdastro", "nl": "Groenbeige"}, {"RAL": 1001, "RGB": "194-176-120", "HEX": "#C2B078", "de": "Beige", "en": "Beige", "fr": "Beige", "es": "Beige", "it": "Beige", "nl": "Beige"}, {"RAL": 1002, "RGB": "198-166-100", "HEX": "#C6A664", "de": "Sand ...
```
What is the source of this? Are we allowed to use this?    
Please add a link/comment

Can you also format this human readable with acceptable line lengths?    
Also please remember to keep the line break at the end of the file.

Lastly, this file shouldn't be located next to the python source code. Move this to another folder.

42Ar replied:
The source is https://gist.github.com/lunohodov/1995178 but the original source is propably https://www.ralfarbpalette.de . I assume that it is legit to use the data.

hd1ex commented on django_plastic_tickets/templates/plastic_tickets/ticket_view.html:
```diff
@@ -43,7 +43,7 @@ <h3> # Ticket {{ ticket.id }}</h3>
                 <td>{{ config.count }}</td>
                 <td>{{ config.material_type }}</td>
-                <td>{{ config.color }}</td>
+                <td>{{ ral_colors|get_item:config.color|get_item:"English" }}</td>
```
This doesn't seem to work for me. I am getting a `None` in the ticket view.

Maybe consider adding a function/property to the model for this rather than using a filter - this should make the view simpler :wink:

hd1ex commented on django_plastic_tickets/templates/plastic_tickets/ticket_view.html:
```diff
@@ -61,4 +61,4 @@ <h3> # Ticket {{ ticket.id }}</h3>
                       wrap="soft">{{ ticket.message }}</textarea>
         </div>
     {% endif %}
-{% endblock %}
```
Please do not remove the newline at the end of files.

hd1ex commented on django_plastic_tickets/views.py:
```diff
@@ -7,10 +7,16 @@
 
 
+@register.filter
+def get_item(dictionary, key):
```
Types are missing here

hd1ex commented on django_plastic_tickets/fixtures/initial_data.json:
```diff
@@ -33,41 +33,64 @@
+            "current_weight": 523,
+            "price": 20,
+            "owner_identifier": "admin",
+            "usage_comment": "Please message the owner about the amount of Filament you have used."
```
I don't like this sample message - this text should rather be in general in some kind of a filament overview.

A better example would be `"This filament does a lot of stringing - use more retraction than usual!"`.

hd1ex reviewed (changes requested):
Thanks for taking care of a model layout to enable proper filament management!

In general I like your ideas about filament management - especially using the RAL color model.    
Can you maybe also extend the description of the [color selection in the file form](https://github.com/hd1ex/django-plastic-tickets/blob/6abc87cde55ac4e21a07442667aa195df5ed2d80/django_plastic_tickets/templates/plastic_tickets/new_ticket_file_form.html#L93) with a reference to RAL?     
Maybe you know a better Website for RAL-Lookup than [Wikipedia](https://en.wikipedia.org/wiki/List_of_RAL_colors) :wink:

I added a few comments below you have to tackle before this is getting merged.    
Also please be informed that this will be the last breaking model change (updating migrations is generally a bad thing but cannot be avoided here) - so make sure this is the base we want in the end :sweat_smile: 

Lastly, here are some comments about the git workflow (maybe read the [contribution guide](https://github.com/hd1ex/plastic-tickets/blob/main/Contributing.md) again):         
- The commit message isn't up to standards and uses a different style than usual ([please read this](https://chris.beams.io/posts/git-commit/)) - also do not forget to update it, when you make further changes in this PR
- Please be prepared to rebase to apply the requested changes ([read this](https://git-rebase.io/), if you are not familiar with this concept)

Don't worry about not getting this right the first time - we are all learning these things by trying out :grin:

Commit:
Restructure database model

- Refine Material fields by specializing Material class
- Drop color table, using RAL color numbers from now on
- Send entered values back to client, so that form is not reset when page reloads
- Refer to material stocks directly in PrintConfig
- Add support for material properties, like transparency

hd1ex commented on django_plastic_tickets/forms.py:
```diff
@@ -17,36 +17,29 @@
         file=active_file, user=user).first()
 
-    # Update existing config and return
+    # Update existing config or create a new
```
`s/or create a new/or create a new one`

hd1ex commented on django_plastic_tickets/forms.py:
```diff
@@ -17,36 +17,29 @@
-                                ticket=None
-                                )
+    else:
+        config = models.PrintConfig(file=active_file,
+                                    count=int(fc),
+                                    user=user,
+                                    ticket=None)
+    config.save()
+    config.material_stocks.set(stocks, clear=True)
```
Can you reason about this change?

42Ar replied:
The PrintConfig object stores now a many to many relationship to the selected materials, because this simplifies the code above (the presence of the material must not be checked by lookup in the db, which would be unnecessary complex now due to the amount of MaterialProperties). Furthermore, the information "Materials which can be used to print the object" is now directly available and must not be calculated with extra code. Viewed externally, the user might select some material type and color but viewed internally, the user selects MaterialStocks, so it makes more sense to reference those objects than to reference the MaterialStocks properties.

hd1ex replied:
:+1:

hd1ex commented on django_plastic_tickets/forms.py:
```diff
@@ -17,36 +17,29 @@
+                                    count=int(fc),
+                                    user=user,
+                                    ticket=None)
+    config.save()
+    config.material_stocks.set(stocks, clear=True)
     config.save()
```
Is this behavior (saving twice) intentional?

42Ar replied:
According https://docs.djangoproject.com/en/3.1/topics/db/examples/many_to_many/ it must be saved (so that it has an id) before a many to many relationship can be set. However, maybe this only applies to new PrintConfig objects, so I have moved the first save into the else branch.

hd1ex commented on django_plastic_tickets/jinja2/plastic_tickets/ticket_mail.txt:
```diff
@@ -3,11 +3,14 @@
 {{ message }}
-{% endif %}
\ No newline at end of file
+{% endif %}
```
Seems like you removed the new line here. Please put it back :D

hd1ex commented on django_plastic_tickets/locale/de/LC_MESSAGES/django.po:
```diff
@@ -18,28 +18,107 @@ msgstr ""
 "Plural-Forms: nplurals=2; plural=(n != 1);\n"
 "X-Generator: Poedit 2.4.1\n"
 
-#: django-plastic-tickets/django_plastic_tickets/forms.py:98
```
These changes look wrong. Let's keep the project root as the origin.

hd1ex commented on django_plastic_tickets/migrations/0002_auto_20210126_2216.py:
```diff
@@ -0,0 +1,34 @@
+# Generated by Django 3.1.2 on 2021-01-26 21:16
```
Please merge all migrations. There is now point in keeping them in this change, if you are redefining the model anyway.
There should only be the (new) initial one.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -42,63 +61,145 @@ class ProductionMethod(DescribedModel):
+        else:
+            return gettext("None")
+
+    def get_disp_list_name(self):
```
Return type is missing.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -119,12 +220,19 @@ def get_message_row_count(self, cols=77) -> int:
+    material_stocks = models.ManyToManyField(MaterialStock)
     ticket = models.ForeignKey(Ticket, on_delete=models.CASCADE, null=True)
 
+    def get_first_material(self):
+        return next(iter(self.material_stocks.all())).material
+
+    def get_color_name(self):
+        return self.get_first_material().get_display_name()
+
+    def get_material_type_name(self):
+        return self.get_first_material().get_disp_list_name()[1]
+
```
Also the return types are missing here.

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -133,3 +241,53 @@ def get_file_url(self):
                 'id': self.ticket.id, 'filename': self.get_file_name()
             }))
+
+    def get_material_stock_list(self):
+        return ", ".join(str(s.label) for s in self.material_stocks.all())
+
+    def __str__(self):
+        if self.ticket is None:
+            tstr = "?"
+        else:
+            tstr = str(self.ticket.pk)
+        return (f'#{tstr}: {self.count}, "{self.file}",'
+                f'{self.get_material_stock_list()}')
```
Types missing

hd1ex commented on django_plastic_tickets/models.py:
```diff
@@ -133,3 +241,53 @@ def get_file_url(self):
+                f'{self.get_material_stock_list()}')
+
+
+def get_material_type_options(stocks: [MaterialStock]):
```
Return type missing.

hd1ex commented on django_plastic_tickets/templates/plastic_tickets/new_ticket_file_form.html:
```diff
@@ -167,10 +167,25 @@ <h4>Advanced information</h4>
+   document.getElementById("reset-btn").onclick = reset;
+   reset();
+
     // TODO: Maybe also display description of (special) material colors
```
This TODO is now done (as we now have special properties) :D

hd1ex commented on django_plastic_tickets/templates/plastic_tickets/ticket_view.html:
```diff
@@ -42,8 +42,8 @@ <h3> # Ticket {{ ticket.id }}</h3>
-                <td>{{ config.material_type }}</td>
-                <td>{{ config.color }}</td>
+                <td>{{ config.get_material_type_name }}</td>
+                <td>{{ config.get_color_name }}</td>
```
Maybe you also want to display the same info as in the ticket email (labels of possible materials) here.
Also consider displaying this only to the staff.

hd1ex reviewed (changes requested):
This change is shaping up really well!

I really like the ideas you brought in and i am looking forward to merge this PR soon.
Here are some comments about the recent changes - most of them are cosmetic ones.

hd1ex reviewed (approved):
LGTM
